lgc/builder/BuilderImpl.cpp

```diff
@@ -130,36 +133,39 @@ Value *BuilderImplBase::CreateIntegerDotProduct(Value *vector1, Value *vector2,
       scalar = CreateAdd(product, scalar);
     }
     if (hasAccumulator) {
-      if (compBitWidth == 8 || compBitWidth == 16) {
-        scalar = CreateTrunc(scalar, getInt32Ty());
-        accumulator = CreateTrunc(accumulator, getInt32Ty());
-        Intrinsic::ID addIntrinsic = isSigned ? Intrinsic::sadd_sat : Intrinsic::uadd_sat;
-        scalar = CreateBinaryIntrinsic(addIntrinsic, scalar, accumulator, nullptr, instName);
-      } else {
-        scalar = CreateAdd(scalar, accumulator);
-      }
+      Intrinsic::ID addIntrinsic = isSigned ? Intrinsic::sadd_sat : Intrinsic::uadd_sat;
+      scalar = CreateBinaryIntrinsic(addIntrinsic, scalar, accumulator, nullptr, instName);
+    } else {
+      scalar = CreateTrunc(scalar, outputTy);
     }
   } else {
+    // <8xi4> is native supported by v_dot8_i32_i4 or v_dot8_i32_u4
     // <4xi8>, <3xi8>, <2xi8> are native supported by using v_dot4_i32_i8 or v_dot4_u32_u8
     // <2xi16> is native supported by using v_dot2_i32_i16 and v_dot2_u32_u16
     // <3xi16>, <4xi16> will be split up with two sequences of v_dot2 and a final saturation add
     Value *input1 = vector1;
     Value *input2 = vector2;
 
-    bool isDot4 = (compBitWidth == 8);
+    bool isDot2 = compBitWidth == 16;
     Value *clamp = hasAccumulator ? getTrue() : getFalse();
 
-    if (isDot4) {
+    if (!isDot2) {
+      assert(compBitWidth == 8 || compBitWidth == 4);
       if (compCount < 4) {
         // Extend <3xi8> or <2xi8> to <4xi8>
         input1 = CreateShuffleVector(input1, Constant::getNullValue(inputTy), ArrayRef<int>({0, 1, 2, 3}));
         input2 = CreateShuffleVector(input2, Constant::getNullValue(inputTy), ArrayRef<int>({0, 1, 2, 3}));
       }
```

Comment:
Should this be updated to handle the case of compBitWidth == 4 and extend to 8 components?

Reply:
My initial understanding is that if the original operands are <3xi8> or <2xi8>, they are expected to be packed as `PackedVectorFormat4x8Bit`. If the original operands are 32-bit integer, they can be interpreted based on the packed vector format.